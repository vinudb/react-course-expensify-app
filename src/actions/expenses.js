import uuid from 'uuid';
import database from '../firebase/firebase';

//component calls action generator
//action generator returns the object
//component dispatches the object
//redux stores the changes


//With firease integration..
//component calls action generator
//action generator returns the function
//component dispatches the function 
//function runs where firebase code resides and also redux stores the changes

//ADD_EXPENSE
export const addExpense = (expense)=>({
    type:'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {})=>{
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = {description, note, amount, createdAt};

        return database.ref('expenses').push(expense).then((ref)=>{
            //dispatch to store in redux
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

//REMOVE_EXPENSE
export const removeExpense = ({id}={})=>({
    type: 'REMOVE_EXPENSE',
    id: id
});

export const startRemoveExpense = ({id}={}) =>{
    return(dispatch)=>{
        return database.ref(`expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense({id}));
        });
    };
};

//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//SET_EXPENSES
export const setExpenses = (expenses)=>({
    type: 'SET_EXPENSES',
    expenses 
});

//fetch all expenses data once
//parse that data in to an array
//dispatch set_expenses, so that data actualyy changes
export const startSetExpenses = ()=>{
    return (dispatch)=>{
        return database.ref('expenses').once('value').then((snapshot)=>{
            const expenses = [];

            snapshot.forEach((childSnapShot)=>{
                expenses.push({
                    id: childSnapShot.key,
                    ...childSnapShot.val()
                });
            });

            dispatch(setExpenses(expenses));
        });
    };

};