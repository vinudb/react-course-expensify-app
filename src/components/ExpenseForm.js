import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
//import 'react-dates/lib/css/_datepicker.css';

// const now = moment();
// console.log(now.format('MMM Do, YYYY'));
//if a component has a form, it will almost always need state. Then the component should be class based component
export default class ExpenseForm extends React.Component{
    constructor(props){
        super(props);
		
        this.state={
            description: props.expense ? props.expense.description:'',
            note:props.expense ? props.expense.note:'',
            amount:props.expense ? (props.expense.amount / 100).toString() :'',
            createdAt: props.expense ? moment(props.expense.createdAt):moment(),
            claenderFocused:false,
            error:''
        };
    }
    onDescriptionChange = (e)=>{
        const description = e.target.value;
        this.setState(()=>({description}));
    };
    onNoteChange=(e)=>{
        const note = e.target.value;
        this.setState(()=>({note}));
    };
    onAmountChange=(e)=>{
        const amount = e.target.value;
        //regex to allow only digits and 2 decimal places //refer regex101 google
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}));
        }
    };
    onDateChange=(createdAt)=>{
        //works only when date is changed and not when date field is deleted from the keyboard
        if(createdAt)
            this.setState(()=>({createdAt}));
    };
    onFocusChange=({focused})=>{
        this.setState(()=>({claenderFocused:focused}));
    };
    
    onSubmit=(e)=>{
        e.preventDefault();
		
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error:'Please type description and amount before submit'}));
            console.log(this.state.error);
        }else{
            this.setState(()=>({error:''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note:this.state.note
            });
        }
    };

    render(){
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.claenderFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1} //to show only one month in the view.
                        isOutsideRange={(day)=>false} //to enable picking previous dates
                    />
                    <textarea
                        placeholder="Add a note for the expense(optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}