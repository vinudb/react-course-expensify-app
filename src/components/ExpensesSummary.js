import React from 'react';
import numeral from 'numeral';
import {connect} from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = ({ expenseCount, expensesTotal })=>{
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formatedExpenseTotal = numeral(expensesTotal/100).format('Rs. 0,00.00');
    return (
        <div>
            <h1>Viewing {expenseCount} {expenseWord} totalling {formatedExpenseTotal} </h1>
        </div>
    );
};

const mapStateToProps = (state) =>{
    const visibleExpense = selectExpenses(state.expenses, state.filters);

    return{
        expenseCount: visibleExpense.length,
        expensesTotal: selectExpensesTotal(visibleExpense)
    };
};

export default connect(mapStateToProps)(ExpenseSummary);