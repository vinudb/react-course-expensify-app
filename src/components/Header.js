import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = ()=>(
    <header>
        <h1>Expensify</h1>
        {/*activeClassName this applies only to the link which is active*/}
        <NavLink to="/" activeClassName="is-active" exact={true}> Dashboard </NavLink>
        <NavLink to="/create" activeClassName="is-active"> Add Expense </NavLink>
        <NavLink to="/help" activeClassName="is-active"> Help </NavLink>
    </header>
);

export default Header;