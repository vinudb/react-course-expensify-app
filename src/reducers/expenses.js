//Expenses reducer
const expensesReducerDefaultState = [];
export default (state=expensesReducerDefaultState, action)=>{
    switch(action.type){
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
        //filter is used to remove an element in the array without changing the original array
            return state.filter(({id})=>action.id !==id);  
        case 'EDIT_EXPENSE':
            return state.map((expense)=>{
                if(expense.id===action.id)
                {
                    return {
                        ...expense,
                        ...action.updates
                    }
                }
                else{
                    return expense;
                }
            });
        case 'SET_EXPENSES':
            return action.expenses;    
        default:
            return state;
    }
};

