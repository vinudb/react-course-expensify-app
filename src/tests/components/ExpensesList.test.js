import React from 'react';
import {shallow} from 'enzyme';
//import toJSON from 'enzyme-to-json';
import {ExpenseList} from '../../components/ExpensesList';
import expenses from '../fixtures/expenses';

test('should render expenselist correctly',()=>{
    const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});