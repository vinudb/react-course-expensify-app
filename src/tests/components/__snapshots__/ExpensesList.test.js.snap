// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenselist correctly 1`] = `
<div>
  <ExpenseListItem
    amount={100}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={10095}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={400}
    createdAt={345600000}
    description="Credit"
    id="3"
    key="3"
    note=""
  />
</div>
`;
